### PR TITLE
chore(docs): document python interfaces

### PR DESCRIPTION
**Pull request by Alex6323**
# Description of change

Fully document Python binding API.

## Links to any relevant issues

#634 

## Type of change

- Documentation Fix

### CODE REVIEW

File: bindings/python/iota_sdk/client/_high_level_api.py [hunk 1]
```diff
@@ -12,18 +12,41 @@
...
         self.start = start
         self.end = end
 
 
 class GenerateAddressOptions():
+    """Options for generating an address.
+    
+    Attributes:
+        internal: whether to generate internal addresses
+        ledgerNanoPrompt: whether to prompt the user for a ledger nano seed
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/client/_high_level_api.py]: What's a ledger nano seed?

Comment (Alex6323) [reply bindings/python/iota_sdk/client/_high_level_api.py]: forgot to fix the AI suggestion

File: bindings/python/iota_sdk/wallet/sync_options.py [hunk 2]
```diff
@@ -34,22 +45,52 @@ def __init__(self,
...
+    account : AccountSyncOptions, optional
+        account sync options
+    alias : AliasSyncOptions, optional
+        alias sync options
+    nft : NftSyncOptions, optional
+        NFT sync options
+    sync_only_most_basic_outputs : bool, optional
+        whether to sync only most basic outputs
+    sync_native_token_foundries : bool, optional
+        whether to sync native token foundries
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/wallet/sync_options.py]: Some of them don't really look helpful, maybe copy the comments from the Rust code?

Comment (Alex6323) [reply bindings/python/iota_sdk/wallet/sync_options.py]: yeah, makes sense

File: bindings/python/iota_sdk/client/_high_level_api.py [hunk 3]
```diff
@@ -56,7 +87,13 @@ def get_outputs(self, output_ids: List[OutputId]) -> List[OutputWithMetadata]:
 
     def get_outputs_ignore_errors(self, output_ids: List[OutputId]) -> List[OutputWithMetadata]:
         """Try to get OutputWithMetadata from provided OutputIds.
-           Requests are sent in parallel and errors are ignored, can be useful for spent outputs.
+        Requests are sent in parallel and errors are ignored, can be useful for spent outputs.
+
+        Args:
+            output_ids (List[OutputId]): list of output ids
```
Comment (qrayven) [inline bindings/python/iota_sdk/client/_high_level_api.py]: Technically (just read from spec) if we use typing, we should omit types in descriptions

File: bindings/python/iota_sdk/client/_node_core_api.py [hunk 4]
```diff
@@ -11,125 +11,161 @@
...
         return self._call_method('postBlockRaw', {
             'blockBytes': block_bytes
         })
 
     def get_output(self, output_id: OutputId) -> OutputWithMetadata:
-        """Get output.
+        """Get the output corresponding to the given output id.
+
+        Returns:
+            OutputWithMetadata: the output itself with its metadata
```
Comment (qrayven) [inline bindings/python/iota_sdk/client/_node_core_api.py]: the rule about typings applies to the `Returns` block too. If you have a type specified, the description should be omitted

File: bindings/python/iota_sdk/client/client.py [hunk 5]
```diff
@@ -49,41 +52,42 @@ def __init__(
...
-        Parameters
-        ----------
-        nodes : string or array of string
-            A single Node URL or an array of URLs.
-        primary_node : string
-            Node which will be tried first for all requests.
-        primary_pow_node : string
-            Node which will be tried first when using remote PoW, even before the primary_node.
-        permanode : string
+        **Arguments**
```
Comment (qrayven) [inline bindings/python/iota_sdk/client/client.py:60]: Just `Args:`  and without "----"  .  
Edit: ( and other occurrences in the code)

File: bindings/python/iota_sdk/types/output.py [hunk 6]
```diff
@@ -22,6 +31,37 @@ class OutputType(IntEnum):
 
 @dataclass
 class Output():
+    """Output.
+
+    **Attributes**
```
Comment (qrayven) [inline bindings/python/iota_sdk/types/output.py:36]: probably without `**` Or is it intentional?

File: bindings/python/iota_sdk/client/client.py (hunk 5 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/client/client.py:60]: I can't use that here. It's limited to maximum of 10 items :cry:

File: bindings/python/iota_sdk/types/output.py (hunk 6 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/types/output.py:36]: Have to use this hack whenever there are 10+ attributes/args unfortunately due to a limitation of `pydoc-markdown` .

File: bindings/python/iota_sdk/client/_high_level_api.py (hunk 3 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/client/_high_level_api.py]: hmm, actually I like removing the types from the docs, yeah. It is supported however and rendered differently which is kinda nice, but still I think we can remove them which is also easier to maintain/keep in sync.

File: bindings/python/iota_sdk/client/_node_core_api.py (hunk 4 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/client/_node_core_api.py]: not the description, but the type you mean, right?

File: bindings/python/iota_sdk/wallet/sync_options.py (hunk 2 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/wallet/sync_options.py]: Done

File: bindings/python/iota_sdk/client/_high_level_api.py (hunk 1 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/client/_high_level_api.py]: Fixed

File: bindings/python/iota_sdk/client/_node_core_api.py (hunk 4 above)
Comment (qrayven) [reply bindings/python/iota_sdk/client/_node_core_api.py]: sorry, yes, the type

File: bindings/python/iota_sdk/client/_high_level_api.py [hunk 7]
```diff
@@ -73,16 +125,30 @@ def find_blocks(self, block_ids: List[HexStr]) -> List[Block]:
...
     def retry(self, block_id: HexStr) -> List[HexStr | Block]:
         """Retries (promotes or reattaches) a block for provided block id. Block should only be
-           retried only if they are valid and haven't been confirmed for a while.
+        retried only if they are valid and haven't been confirmed for a while.
+
+        Args:
+            block_id: A block id.
+
+        Returns:
+            A list of `HexStr`s or `Block`s.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/client/_high_level_api.py]: It's a list where the first element is the block id and the second one the block

File: bindings/python/iota_sdk/client/_high_level_api.py [hunk 8]
```diff
@@ -73,16 +125,30 @@ def find_blocks(self, block_ids: List[HexStr]) -> List[Block]:
...
+        milestone). Default interval is 5 seconds and max attempts is 40. Returns the included block at first
+        position and additional reattached blocks.
+
+        Args:
+            block_id: A block id.
+            interval: A retry interval in seconds. Defaults to 5.
+            max_attempts: A maximum number of retries. Defaults to 40.
+
+        Returns:
+            A list of lists of `HexStr`s or `Block`s.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/client/_high_level_api.py]: And here it's a List of Lists where the first element is the block id and the second one the block

File: bindings/python/iota_sdk/client/_high_level_api.py [hunk 9]
```diff
@@ -116,16 +193,26 @@ def find_inputs(self, addresses: List[str], amount: int):
...
-        """Reattaches blocks for provided block id. Blocks can be reattached only if they are valid and haven't been
-           confirmed for a while.
+        """Reattaches blocks for a provided block id. Blocks can be reattached only if they are valid and
+        haven't been confirmed for a while.
+
+        Args:
+            block_id: A block id of a block that should be reattached.
+
+        Returns:
+            The list of reattached blocks.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/client/_high_level_api.py]:
```suggestion
            The reattached block id and block.
```

File: bindings/python/iota_sdk/client/_high_level_api.py [hunk 10]
```diff
@@ -135,6 +222,12 @@ def reattach(self, block_id: HexStr) -> List[HexStr | Block]:
 
     def reattach_unchecked(self, block_id: HexStr) -> List[HexStr | Block]:
         """Reattach a block without checking if it should be reattached.
+
+        Args:
+            block_id: A block id of a block that should be reattached.
+
+        Returns:
+            The list of reattached blocks.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/client/_high_level_api.py]:
```suggestion
            The reattached block id and block.
```

File: bindings/python/iota_sdk/client/_high_level_api.py [hunk 11]
```diff
@@ -143,8 +236,14 @@ def reattach_unchecked(self, block_id: HexStr) -> List[HexStr | Block]:
...
-        """Promotes a block. The method should validate if a promotion is necessary through get_block. If not, the
-           method should error out and should not allow unnecessary promotions.
+        """Promotes a block. The method should validate if a promotion is necessary through get_block.
+        If not, the method should error out and should not allow unnecessary promotions.
+
+        Args:
+            block_id: A block id of a block that should be promoted.
+
+        Returns:
+            The list of promoted blocks.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/client/_high_level_api.py]:
```suggestion
            The block id and block that promoted the provided block.
```

File: bindings/python/iota_sdk/client/_high_level_api.py [hunk 12]
```diff
@@ -154,6 +253,12 @@ def promote(self, block_id: HexStr) -> List[HexStr | Block]:
 
     def promote_unchecked(self, block_id: HexStr) -> List[HexStr | Block]:
         """Promote a block without checking if it should be promoted.
+
+        Args:
+            block_id: A block id of a block that should be promoted.
+
+        Returns:
+            The list of promoted blocks.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/client/_high_level_api.py]:
```suggestion
            The block id and block that promoted the provided block.
```

File: bindings/python/iota_sdk/client/client.py [hunk 13]
```diff
@@ -164,30 +175,19 @@ def build_alias_output(self,
...
+        Args:
+            alias_id: TODO.
+            unlock_conditions: TODO.
+            amount: TODO.
+            native_tokens: TODO.
+            state_index: TODO.
+            state_metadata: TODO.
+            foundry_counter: TODO.
+            features: TODO.
+            immutable_features: TODO.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/client/client.py]: wen

File: bindings/python/iota_sdk/client/client.py [hunk 14]
```diff
@@ -225,20 +225,14 @@ def build_basic_output(self,
...
-            Features for this outputs
-
-        Returns
-        -------
-        Output as dict
+        Args:
+            unlock_conditions: The unlock conditions for this output.
+            amount: TODO.
+            native_tokens: TODO.
+            features: TODO.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/client/client.py]: wen

File: bindings/python/iota_sdk/client/client.py [hunk 15]
```diff
@@ -388,36 +365,23 @@ def build_and_post_block(self,
...
-        tag : string
-            Hex encoded tag
-
-        Returns
-        -------
-        Block as dict
-
+        **Arguments**
+        account_index : TODO.
+        coin_type : TODO.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/client/client.py]: wen

File: bindings/python/iota_sdk/types/common.py [hunk 16]
```diff
@@ -19,21 +26,18 @@ def __int__(self):
...
-        disabled : bool
-            Disable node
+        """Initialize a Node.
+
+        Args:
+            url: The node url.
+            jwt: A JWT token for authentication.
+            username: A username for basic authentication.
+            password: A password for basic authentication.
+            disabled: TODO.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/types/common.py]:
```suggestion
            disabled: Whether the node should be used for API requests or not.
```

File: bindings/python/iota_sdk/types/output.py [hunk 17]
```diff
@@ -22,6 +31,37 @@ class OutputType(IntEnum):
...
+        Foundry counter
+    features : List[Feature], optional
+        Features
+    nativeTokens : List[NativeToken], optional
+        Native tokens
+    immutableFeatures : List[Feature], optional
+        Immutable features
+    serialNumber : int, optional 
+        Serial number
+    tokenScheme : TokenScheme, optional
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/types/output.py]: Remove the types?

File: bindings/python/iota_sdk/types/output.py [hunk 18]
```diff
@@ -59,6 +99,18 @@ def as_dict(self):
...
+        blockId (HexStr): the corresponding block ID in which the output appeared
+        transactionId (HexStr): the transaction ID where that output was created
+        outputIndex (int): the index of the output within the corresponding transaction
+        isSpent (bool): whether the output is spent
+        milestoneIndexBooked (int): the milestone index where that output was created
+        milestoneTimestampBooked (int): the milestone timestamp where that output was created
+        ledgerIndex (int): the ledger index at which the output appeared
+        milestoneIndexSpent (int, optional): the milestone index spent where that output was spent
+        milestoneTimestampSpent (int, optional): the milestone timestamp spent where that output was spent
+        transactionIdSpent (HexStr, optional): the transaction ID that spent the output
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/types/output.py]: Also remove the types?

File: bindings/python/iota_sdk/types/output_data.py [hunk 19]
```diff
@@ -11,6 +11,19 @@
...
+
+    Attributes:
+        outputId (HexStr): the associated output ID
+        metadata (OutputMetadata): metadata associated with the output
+        output (Output): the output itself
+        isSpent (bool): whether the output is spent
+        address (Address): the address associated with the output
+        networkId (str): the network ID associated with the output
+        remainder (bool): whether the output is a remainder
+        chain (List[int], optional): array of chain state indexes
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/types/output_data.py]: types

File: bindings/python/iota_sdk/types/output_id.py [hunk 20]
```diff
@@ -5,6 +5,14 @@
 
 
 class OutputId(dict):
+    """Represents an output ID.
+
+    Attributes:
+        output_id (HexStr): The unique id of an output.
+        transaction_id (HexStr): The transaction id associated with the output.
+        output_index (int): The index of the output within the transaction.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/types/output_id.py]: types

File: bindings/python/iota_sdk/types/output_id.py [hunk 21]
```diff
@@ -24,6 +32,14 @@ def __init__(self, transaction_id: HexStr, output_index: int):
 
     @classmethod
     def from_string(cls, output_id: HexStr):
+        """Creates an `OutputId` instance from a `HexStr`.
+
+        Args:
+            output_id (HexStr): The unique id of an output as a hex string.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/types/output_id.py]:
```suggestion
            output_id: The unique id of an output as a hex string.
```

File: bindings/python/iota_sdk/types/payload.py [hunk 22]
```diff
@@ -10,15 +10,40 @@
...
 class Payload():
-    def __init__(self, type,  milestone: Optional[Any] = None, tagged_data=None, transaction=None, treasury_transaction: Optional[Any] = None):
+    """Base class for block payloads.
+
+    Attributes:
+        type (int): Payload type
+        milestone (optional): Milestone payload
+        tagged_data (optional): TaggedData payload
+        transaction (optional): Transaction payload
+        treasury_transaction (optional): TreasuryTransaction payload
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/types/payload.py]: types

File: bindings/python/iota_sdk/types/payload.py [hunk 23]
```diff
@@ -10,15 +10,40 @@
...
+    """
+    def __init__(self, type, milestone=None, tagged_data=None, transaction=None, treasury_transaction=None):
         """Initialize a payload
+
+        Args:
+            type (int): Payload type
+            milestone (optional): Milestone payload
+            tagged_data (optional): TaggedData payload
+            transaction (optional): Transaction payload
+            treasury_transaction (optional): TreasuryTransaction payload
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/types/payload.py]: types

File: bindings/python/iota_sdk/types/payload.py [hunk 24]
```diff
@@ -45,7 +70,19 @@ def as_dict(self):
...
+        index (int): corresponding milestone index
+        timestamp (int): corresponding milestone timestamp
+        protocolVersion (int): current protocol version
+        previousMilestoneId (HexStr): previous milestone id
+        parents (List[HexStr]): parents of the milestone 
+        inclusionMerkleRoot (HexStr): the inclusion merkle root
+        appliedMerkleRoot (HexStr): the applied merkle root
+        signatures (List[Ed25519Signature]): milestone signatures
+        options (List[Any], optional): milestone options
+        metadata (HexStr, optional): milestone metadata
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/types/payload.py]: types

Comment (Thoralf-M) [reply bindings/python/iota_sdk/types/payload.py]: More below in other files, will not comment there now

File: bindings/python/iota_sdk/client/client.py (hunk 13 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/client/client.py]: if good suggestion then immediatedly :joy:

File: bindings/python/iota_sdk/client/_high_level_api.py (hunk 3 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/client/_high_level_api.py]: Should be done unless I forget something.

File: bindings/python/iota_sdk/client/_high_level_api.py (hunk 7 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/client/_high_level_api.py]: now I get it ... I was so confused :joy:

File: bindings/python/iota_sdk/types/output.py (hunk 17 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/types/output.py]: yeah ... that went into `part 2` (see previous commit message :stuck_out_tongue_closed_eyes: )

File: bindings/python/iota_sdk/types/output.py (hunk 18 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/types/output.py]: yeah ... that went into `part 2` (see previous commit message :stuck_out_tongue_closed_eyes: )

File: bindings/python/iota_sdk/types/output_data.py (hunk 19 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/types/output_data.py]: yeah ... that went into `part 2` (see previous commit message :stuck_out_tongue_closed_eyes: )

File: bindings/python/iota_sdk/types/output_id.py (hunk 20 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/types/output_id.py]: yeah ... that went into `part 2` (see previous commit message :stuck_out_tongue_closed_eyes: )

File: bindings/python/iota_sdk/types/payload.py (hunk 22 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/types/payload.py]: yeah ... that went into `part 2` (see previous commit message :stuck_out_tongue_closed_eyes: )

File: bindings/python/iota_sdk/types/payload.py (hunk 23 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/types/payload.py]: yeah ... that went into `part 2` (see previous commit message :stuck_out_tongue_closed_eyes: )

File: bindings/python/iota_sdk/types/payload.py (hunk 24 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/types/payload.py]: yeah ... that went into `part 2` (see previous commit message :stuck_out_tongue_closed_eyes: )

File: bindings/python/iota_sdk/client/_high_level_api.py (hunk 7 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/client/_high_level_api.py]: Fixed

File: bindings/python/iota_sdk/client/_high_level_api.py (hunk 8 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/client/_high_level_api.py]: Fixed

File: bindings/python/iota_sdk/client/client.py (hunk 13 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/client/client.py]: Done

File: bindings/python/iota_sdk/client/client.py (hunk 15 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/client/client.py]: now

File: bindings/python/iota_sdk/client/_node_indexer_api.py [hunk 25]
```diff
@@ -110,7 +191,10 @@ def foundry_output_ids(
         return self.OutputIdsResponse(response)
 
     def foundry_output_id(self, foundry_id: HexStr) -> OutputId:
-        """Fetch foundry Output ID.
+        """Fetch foundry Output ID from the given query parameters.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/client/_node_indexer_api.py]: No query parameters here

File: bindings/python/iota_sdk/client/_node_indexer_api.py [hunk 26]
```diff
@@ -91,15 +166,21 @@ def nft_output_ids(
         return self.OutputIdsResponse(response)
 
     def nft_output_id(self, nft_id: HexStr) -> OutputId:
-        """Fetch NFT output ID.
+        """Fetch NFT output ID from the given query parameters.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/client/_node_indexer_api.py]: No query parameters here

File: bindings/python/iota_sdk/client/_node_indexer_api.py [hunk 27]
```diff
@@ -72,15 +141,21 @@ def alias_output_ids(
         return self.OutputIdsResponse(response)
 
     def alias_output_id(self, alias_id: HexStr) -> OutputId:
-        """Fetch alias output ID.
+        """Fetch alias output ID from the given query parameters.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/client/_node_indexer_api.py]: No query parameters here

File: bindings/python/iota_sdk/client/_node_indexer_api.py (hunk 25 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/client/_node_indexer_api.py]: :see_no_evil:

File: bindings/python/iota_sdk/client/_node_indexer_api.py (hunk 26 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/client/_node_indexer_api.py]: :see_no_evil:

File: bindings/python/iota_sdk/client/_node_indexer_api.py (hunk 27 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/client/_node_indexer_api.py]: :see_no_evil:

File: bindings/python/iota_sdk/types/signature.py [hunk 28]
```diff
@@ -7,13 +7,28 @@
 
 @dataclass
 class Ed25519Signature():
+    """An Ed25519 signature.
+
+    Attributes:
+        publicKey: The Ed25519 public key.
+        signature: The Ed25519 signature of some message.
+        type: The Ed25519 signature type code.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/types/signature.py]:
```suggestion
        type: The Ed25519 signature type.
```

File: bindings/python/iota_sdk/types/unlock_condition.py [hunk 29]
```diff
@@ -38,90 +59,95 @@ def as_dict(self):
...
-        Parameters
-        ----------
-        amount : int
-            Amount
-        return_address : Address
-            Return address
+        """Initialize `Self`.
+
+        Args:
+            amount: The amount of base coins the consuming transaction should deposit to `return_address`.
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/types/unlock_condition.py]:
```suggestion
            amount: The amount of base coins the consuming transaction must deposit to `return_address`.
```

**Review by Thoralf-M — DISMISSED** [review]

**Review by qrayven — DISMISSED** [review]

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
CI failing

File: bindings/python/iota_sdk/client/client.py [hunk 30]
```diff
@@ -467,10 +467,7 @@ def unhealthy_nodes(self) -> List[Dict[str, Any]]:
         """
         return self._call_method('unhealthyNodes')
 
-    def prepare_transaction(self,
-                            secret_manager: Optional[LedgerNanoSecretManager | MnemonicSecretManager |
-                                                     SeedSecretManager | StrongholdSecretManager] = None,
-                            options: Optional[TransactionOptions] = None):
+    def prepare_transaction(self, secret_manager = None, options = None):
```
Comment (Thoralf-M) [inline bindings/python/iota_sdk/client/client.py]: No no no

Comment (Thoralf-M) [reply bindings/python/iota_sdk/client/client.py]: https://github.com/Alex6323/iota-sdk/pull/1

**Review by Thoralf-M — DISMISSED** [review]

**Review by qrayven — DISMISSED** [review]

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Conflicts

**Review by Thoralf-M — APPROVED** [review]

**Review by qrayven — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**